torch/csrc/api/include/torch/data/dataloader.h

```diff
@@ -65,228 +52,65 @@ class DataLoader {
-  }
-
-  /// Returns the next batch of data, or an empty `optional` if the `DataLoader`
-  /// is exhausted. This operation will block until a batch is available.
-  optional<Batch> next() {
-    optional<Batch> batch;
-    if (options_.workers > 0) {
-      optional<Result> result = sequencer_->next(
-          [this] { return this->shuttle_.pop_result(this->options_.timeout); });
-      if (result) {
-        if (result->exception) {
-          throw WorkerException(result->exception);
-        } else {
-          AT_ASSERT(result->batch.has_value());
-          batch = std::move(result->batch);
-          prefetch(1);
-        }
-      }
-    } else if (auto batch_request = get_batch_request()) {
-      AT_ASSERT(main_thread_dataset_ != nullptr);
-      batch = main_thread_dataset_->get_batch(std::move(*batch_request));
-    }
-    return batch;
-  }
-
-  /// The function that worker threads run.
-  void worker_thread(Dataset dataset) {
-    while (true) {
-      auto job = shuttle_.pop_job();
-      if (job.quit) {
-        break;
-      }
-      try {
-        auto batch = dataset.get_batch(std::move(*job.batch_request));
-        shuttle_.push_result({std::move(batch), job.sequence_number});
-      } catch (...) {
-        shuttle_.push_result({std::current_exception(), job.sequence_number});
-      }
-    }
-  }
-
-  optional<BatchRequest> get_batch_request() {
-    auto indices = sampler_.next(options_.batch_size);
-    if (!indices ||
-        (indices->size() < options_.batch_size && options_.drop_last)) {
-      return nullopt;
-    }
-    AT_ASSERT(indices->size() > 0);
-    return indices;
-  }
-
-  template <typename T>
-  void push_job(T value) {
-    shuttle_.push_job({std::move(value), sequence_number_++});
-  }
-
-  std::unique_ptr<detail::sequencers::Sequencer<Result>> new_sequencer() {
-    if (options_.enforce_ordering) {
-      return torch::make_unique<detail::sequencers::OrderedSequencer<Result>>(
-          options_.max_jobs);
-    }
-    return torch::make_unique<detail::sequencers::NoSequencer<Result>>();
-  }
-
-  /// The options the `DataLoader` was configured with.
-  const FullDataLoaderOptions options_;
-
-  /// The dataset for the main thread, only has a value if the number of
-  /// worker threads was configured as zero, meaning the main thread has to do
-  /// all the work (synchronously). NOTE: Really want this to be on the heap
-  /// when empty, therefore `unique_ptr` and not `optional`.
-  std::unique_ptr<Dataset> main_thread_dataset_;
-
-  /// The sampler with which new batch requests are created.
-  Sampler sampler_;
-
-  /// The sequence number for the *next* batch to be retrieved from the
-  /// dataset.
-  size_t sequence_number_ = 0;
-
-  /// The worker threads, running the `worker_thread()` method.
-  std::vector<std::thread> workers_;
-
-  /// The `DataShuttle` which takes care of the life cycle of a job.
-  detail::DataShuttle<Job, Result> shuttle_;
+  Impl impl_;
+}; // namespace data
```

Comment:
Do we really even need this class now? It seems like we could just use the `Impl` types in place of `DataLoader<Impl>` and we wouldn't loose much.

Reply:
That's a great point